homeassistant/components/bluetooth/manager.py

```diff
@@ -483,15 +483,15 @@ def _async_remove_callback() -> None:
         # immediately with the last packet so the subscriber can see the
         # device.
         all_history = self._get_history_by_type(connectable)
-        if (
-            (address := callback_matcher.get(ADDRESS))
-            and (service_info := all_history.get(address))
-            and ble_device_matches(callback_matcher, service_info)
-        ):
-            try:
-                callback(service_info, BluetoothChange.ADVERTISEMENT)
-            except Exception:  # pylint: disable=broad-except
-                _LOGGER.exception("Error in bluetooth callback")
+        for service_info in all_history.values():
+            address_match = callback_matcher.get(ADDRESS)
+            if (
+                address_match is None or address_match == service_info.address
+            ) and ble_device_matches(callback_matcher, service_info):
```

Comment:
It is a bit silly that ble_device_matcher ignores address. Maybe we should re-visit that choice?

Reply:
This was intentional so it can be used for discovery and the callback matcher

Reply:
And.. if this hadn't been the case we might have forgotten the optimization. So probably a good idea :)